src/main/java/org/sonarsource/scanner/maven/bootstrap/JavaVersionResolver.java

```diff
@@ -0,0 +1,114 @@
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+public class JavaVersionResolver {
+  private static final String MAVEN_COMPILER_PLUGIN = "maven-compiler-plugin";
+  private static final String COMPILE_GOAL = MavenUtils.GROUP_ID_APACHE_MAVEN + ":" + MAVEN_COMPILER_PLUGIN + ":compile";
+
+  private final MavenSession session;
+  private final LifecycleExecutor lifecycleExecutor;
+  private final BuildPluginManager buildPluginManager;
+  private final Log log;
+
+  public JavaVersionResolver(MavenSession session, LifecycleExecutor lifecycleExecutor, BuildPluginManager buildPluginManager, Log log) {
+    this.session = session;
+    this.lifecycleExecutor = lifecycleExecutor;
+    this.buildPluginManager = buildPluginManager;
+    this.log = log;
+  }
+
+  /**
+   * Returns the version of Java used by the maven compiler plugin
+   *
+   * @param pom the project pom
+   * @return the java version
+   */
+  @CheckForNull
+  public String getTarget(MavenProject pom) {
+    return coalesce(getString(COMPILE_GOAL, "target"), MavenUtils.getPluginSetting(pom, MavenUtils.GROUP_ID_APACHE_MAVEN, MAVEN_COMPILER_PLUGIN, "target", null));
+  }
+
+  @CheckForNull
+  public String getSource(MavenProject pom) {
+    return coalesce(getString(COMPILE_GOAL, "source"), MavenUtils.getPluginSetting(pom, MavenUtils.GROUP_ID_APACHE_MAVEN, MAVEN_COMPILER_PLUGIN, "source", null));
+
+  }
+
+  private <T> T coalesce(T... objs) {
+    for (T o : objs) {
+      if (o != null) {
+        return o;
+      }
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private String getString(String goal, String parameter) {
+    try {
+      List<MojoExecution> mojoExecutions = lifecycleExecutor.calculateExecutionPlan(session, true, goal).getMojoExecutions();
+      for (MojoExecution exec : mojoExecutions) {
+        Xpp3Dom configuration = exec.getConfiguration();
+        PlexusConfiguration pomConfiguration = new XmlPlexusConfiguration(configuration);
+        PlexusConfiguration config = pomConfiguration.getChild(parameter);
+
+        ConverterLookup converterLookup = new DefaultConverterLookup();
+        MojoDescriptor mojoDescriptor = exec.getMojoDescriptor();
+        PluginDescriptor pluginDescriptor = mojoDescriptor.getPluginDescriptor();
+        ClassRealm pluginRealm = buildPluginManager.getPluginRealm(session, pluginDescriptor);
+        PluginParameterExpressionEvaluator expressionEvaluator = new PluginParameterExpressionEvaluator(session, exec);
+        ConfigurationConverter converter = new StringConverter();
+
+        String value = (String) converter.fromConfiguration(converterLookup, config, String.class, mojoDescriptor.getImplementationClass(),
+          pluginRealm, expressionEvaluator);
+        if (value != null) {
+          return value;
+        }
+      }
+    } catch (Exception e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1166)